kernel/src/hil/hmac.rs

```diff
@@ -0,0 +1,45 @@
+//! Interface for HMAC
+
+use crate::mem::AppSlice;
+use crate::mem::Shared;
+use crate::returncode::ReturnCode;
+
+#[derive(PartialEq)]
+pub enum HMACHash {
+    Sha256,
+}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    // On success the `result` will contain the digest. On error it will contain
+    // the HMAC error value.
+    fn hash_done(&'a self, result: u32);
+}
+
+pub trait HMAC<'a> {
+    /// Set the hash algorithm and the key size
+    fn set_hash(&'a self, hash: HMACHash, key_size: usize) -> ReturnCode;
+
+    /// Set the client instance which will receive `hash_done()` callbacks
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Set the encryption key.
+    /// Returns `EINVAL` if length is not `HMAC_KEY_SIZE`
+    fn set_key(&self, key: &AppSlice<Shared, u8>) -> ReturnCode;
```

Comment:
I don't think `AppSlice` should be showing up in HIL traits. It implies you can't call it from within the kernel. 

As for a model for how a single struct can implement multiple related-but-distinct algorithms in a general way, I'd suggest looking at symmetric_encyrption; note that it does have different modes but does have a single trait for a particular block width.

Reply:
Ok, good point.

If not `AppSlice` then I think the only option is a raw pointer, or is there something else we could use? I was trying to avoid unsafe code, but I don't see a way around it.

Reply:
As for the algorithm, I was expecting that the caller would set this.

For example one app would do something like this (via exposed syscalls):
```rust
set_hash(MyHash);
set_client(callback);
set_key(MyKey);
add_data(MyData1);
add_data(MyData2);
run()
get_digest(&MyDigest);
```

then another app can do the same thing, for it's own has and data. The hmac capsule will enqueue everything until run and should enforce the separation of the apps.

At the moment there is no syscall to specify the hash, I will add that.

Reply:
I pushed an update not using `AppSlice`, let me know what you think